src/btree/bt_sync.c

```diff
@@ -122,8 +122,15 @@ __sync_ref_is_obsolete(WT_SESSION_IMPL *session, WT_REF *ref)
         return (false);
     }
 
+    /* Ignore deleted pages. */
+    if (ref->state == WT_REF_DELETED) {
+        __wt_verbose(session, WT_VERB_CHECKPOINT_GC, "%p: skipping deleted page", (void *)ref);
+        return (false);
+    }
+
     /* Ignore internal pages, these are taken care of during reconciliation. */
-    if (!__wt_ref_is_leaf(session, ref)) {
+    if ((ref->addr != NULL && !__wt_ref_is_leaf(session, ref)) ||
+      (ref->page != NULL && WT_PAGE_IS_INTERNAL(ref->page))) {
```

Comment:
This doesn't look safe to me.

If this code is called from `__sync_ref_int_obsolete_cleanup()`, what prevents the `WT_REF` from being evicted out from under us? That is, can't `WT_ref.page` be switched to NULL after the test and before the call to `WT_PAGE_IS_INTERNAL`?

Maybe based on where we are in the checkpoint walk at that point it can't happen, is there a guarantee I'm missing?

Reply:
Yes, you are correct. I don't think this check is required. Anyway, when the leaf page is clean, it is handled in the different code path. I will remove this code.